cranelift-codegen/src/isa/x86/fde.rs

```diff
@@ -0,0 +1,354 @@
+//! Support for FDE data generation.
+
+use crate::binemit::{FrameUnwindOffset, FrameUnwindSink, Reloc};
+use crate::ir::{FrameLayoutChange, Function};
+use crate::isa::{CallConv, RegUnit, TargetIsa};
+use alloc::vec::Vec;
+use core::convert::TryInto;
+use gimli::write::{
+    Address, CallFrameInstruction, CommonInformationEntry, EhFrame, EndianVec,
+    FrameDescriptionEntry, FrameTable, Result, Writer,
+};
+use gimli::{Encoding, Format, LittleEndian, Register, X86_64};
+use std::ptr;
+
+pub type FDERelocEntry = (FrameUnwindOffset, Reloc);
+
+const FUNCTION_ENTRY_ADDRESS: Address = Address::Symbol {
+    symbol: 0,
+    addend: 0,
+};
+
+#[derive(Clone)]
+struct FDEWriter {
+    vec: EndianVec<LittleEndian>,
+    relocs: Vec<FDERelocEntry>,
+}
+
+impl FDEWriter {
+    fn new() -> Self {
+        Self {
+            vec: EndianVec::new(LittleEndian),
+            relocs: Vec::new(),
+        }
+    }
+    fn into_vec_and_relocs(self) -> (Vec<u8>, Vec<FDERelocEntry>) {
+        (self.vec.into_vec(), self.relocs)
+    }
+}
+
+impl Writer for FDEWriter {
+    type Endian = LittleEndian;
+    fn endian(&self) -> Self::Endian {
+        LittleEndian
+    }
+    fn len(&self) -> usize {
+        self.vec.len()
+    }
+    fn write(&mut self, bytes: &[u8]) -> Result<()> {
+        self.vec.write(bytes)
+    }
+    fn write_at(&mut self, offset: usize, bytes: &[u8]) -> Result<()> {
+        self.vec.write_at(offset, bytes)
+    }
+    fn write_address(&mut self, address: Address, size: u8) -> Result<()> {
+        match address {
+            Address::Constant(_) => self.vec.write_address(address, size),
+            Address::Symbol { .. } => {
+                assert_eq!(address, FUNCTION_ENTRY_ADDRESS);
+                let rt = match size {
+                    4 => Reloc::Abs4,
+                    8 => Reloc::Abs8,
```

Comment:
I don't think 8-byte absolute relocations are necessarily available for all targets - specifically, as far as I know, MachO objects only permit up to dword-size relocations. This is one of the reasons I used the address encoding `SDATA4` in #902. Are qword relocations available in MachO targets? If so, I also prefer absolute relocations, but I'd like to not exclude macos here.

Reply:
This serialization targets specific use case of libunwind (in context of JIT), and it requires FDE in specific format. This PR will not address other use cases.

Reply:
good to know, thank you - I'll be revising #902 to support writing .eh_frame to object files, and I'll make sure to preserve this relocation style.